File: app/entity/contract.go
package entity

import (
	"time"

	"github.com/music-gang/music-gang-api/app/apperr"
)

// Visibility consts for the visibility of the contract.
const (
	VisibilityPrivate = "private"
	VisibilityPublic  = "public"
)

// Visibility defines the visibility of a contract.
type Visibility string

// Validate validates the visibility.
func (v Visibility) Validate() error {
	switch v {
	case
		VisibilityPrivate,
		VisibilityPublic:
		return nil
	default:
		return apperr.Errorf(apperr.EINVALID, "invalid visibility")
	}
}

// Contracts represents a list of contracts.
type Contracts []*Contract

// Contract represents a contract.
// The contract is a cloud function that is executed on a server, deployed by users;
// The contract can have multiple revisions.
type Contract struct {
	ID          int64      `json:"id"`
	Name        string     `json:"name"`
	Description string     `json:"description"`
	UserID      int64      `json:"user_id"`
	Visibility  Visibility `json:"visibility"`
	MaxFuel     Fuel       `json:"max_fuel"` // The maximum amount of fuel that can be burned from the contract.
	Stateful    bool       `json:"stateful"` // Enables the contract to persist its state during different executions (of same revision).
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`

	// avoid to access this field directly because it can be nil, use the Revision method UnwrapRevision instead.
	LastRevision *Revision `json:"last_revision"`
	User         *User     `json:"user"`
}

// MaxExecutionTime returns the maximum execution time of the contract.
// MaxExecutionTime is based on max fuel compared with fuelAmountTable.
func (c *Contract) MaxExecutionTime() time.Duration {
	return MaxExecutionTimeFromFuel(c.MaxFuel)
}

// Validate validates the contract.
func (c *Contract) Validate() error {

	if c.Name == "" {
		return apperr.Errorf(apperr.EINVALID, "contract name is required")
	}

	if c.UserID == 0 {
		return apperr.Errorf(apperr.EINVALID, "User ID is required")
	}

	if c.MaxFuel == 0 {
		return apperr.Errorf(apperr.EINVALID, "Max fuel is required")
	}

	if err := c.Visibility.Validate(); err != nil {
		return err
	}

	return nil
}

// UnwrapRevision returns the last revision of the contract if it exists, otherwise error is returned.
func (c *Contract) UnwrapRevision() (*Revision, error) {
	if c.LastRevision == nil {
		return nil, apperr.Errorf(apperr.ENOTFOUND, "no revision found")
	}
	return c.LastRevision, nil
}

File: tests/entity/contract.rb
# frozen_string_literal: true

# Contract entity.
class Contract
  include Jsonizable

  attr_accessor :id, :name, :description, :user_id, :visibility, :max_fuel, :stateful, :created_at, :updated_at

  def initialize(id: nil,
                 name: nil,
                 description: nil,
                 user_id: nil,
                 visibility: nil,
                 max_fuel: nil,
                 stateful: false,
                 created_at: nil,
                 updated_at: nil)
    @id = id
    @name = name
    @description = description
    @user_id = user_id
    @visibility = visibility
    @max_fuel = max_fuel
    @stateful = stateful
    @created_at = created_at
    @updated_at = updated_at
  end

  def to_hash
    {
      id: @id,
      name: @name,
      description: @description,
      user_id: @user_id,
      visibility: @visibility,
      max_fuel: @max_fuel,
      stateful: @stateful,
      created_at: @created_at,
      updated_at: @updated_at
    }
  end

  class << self
    def from_hash(hash)
      validate_hash hash

      Contract.new id: hash[:id], name: hash[:name], description: hash[:description], user_id: hash[:user_id], visibility: hash[:visibility], max_fuel: hash[:max_fuel], stateful: hash[:stateful], created_at: Time.parse(hash[:created_at]), updated_at: Time.parse(hash[:updated_at])
    end

    def validate_hash(hash)
      raise 'missing id' unless hash.key? :id
      raise 'missing name' unless hash.key? :name
      raise 'missing description' unless hash.key? :description
      raise 'missing user id' unless hash.key? :user_id
      raise 'missing visibility' unless hash.key? :visibility
      raise 'missing max fuel' unless hash.key? :max_fuel
      raise 'missing stateful' unless hash.key? :stateful
      raise 'missing created at' unless hash.key? :created_at
      raise 'missing updated at' unless hash.key? :updated_at
    end
  end
end
